File: mlflow/entities/span_event.py
import json
import sys
import time
import traceback
from dataclasses import dataclass, field
from datetime import datetime

from opentelemetry.util.types import AttributeValue

from mlflow.entities._mlflow_object import _MlflowObject
from mlflow.protos.databricks_trace_server_pb2 import Span as ProtoSpan
from mlflow.utils.proto_json_utils import set_pb_value


@dataclass
class SpanEvent(_MlflowObject):
    """
    An event that records a specific occurrences or moments in time
    during a span, such as an exception being thrown. Compatible with OpenTelemetry.

    Args:
        name: Name of the event.
        timestamp:  The exact time the event occurred, measured in microseconds.
            If not provided, the current time will be used.
        attributes: A collection of key-value pairs representing detailed
            attributes of the event, such as the exception stack trace.
            Attributes value must be one of ``[str, int, float, bool, bytes]``
            or a sequence of these types.
    """

    name: str
    # Use current time if not provided. We need to use default factory otherwise
    # the default value will be fixed to the build time of the class.
    timestamp: int = field(default_factory=lambda: int(time.time() * 1e6))
    attributes: dict[str, AttributeValue] = field(default_factory=dict)

    @classmethod
    def from_exception(cls, exception: Exception):
        "Create a span event from an exception."

        stack_trace = cls._get_stacktrace(exception)
        return cls(
            name="exception",
            attributes={
                "exception.message": str(exception),
                "exception.type": exception.__class__.__name__,
                "exception.stacktrace": stack_trace,
            },
        )

    @staticmethod
    def _get_stacktrace(error: BaseException) -> str:
        """Get the stacktrace of the parent error."""
        msg = repr(error)
        try:
            if sys.version_info < (3, 10):
                tb = traceback.format_exception(error.__class__, error, error.__traceback__)
            else:
                tb = traceback.format_exception(error)
            return "".join(tb).strip()
        except Exception:
            return msg

    def json(self):
        return {
            "name": self.name,
            "timestamp": self.timestamp,
            "attributes": json.dumps(self.attributes, cls=CustomEncoder)
            if self.attributes
            else None,
        }

    def to_proto(self):
        """Convert into OTLP compatible proto object to sent to the Databricks Trace Server."""
        proto = ProtoSpan.Event(
            name=self.name,
            time_unix_nano=self.timestamp,
        )
        # Trace server's proto uses map<string, google.protobuf.Value> for attributes
        for key, value in self.attributes.items():
            set_pb_value(proto.attributes[key], value)
        return proto


class CustomEncoder(json.JSONEncoder):
    """
    Custom encoder to handle json serialization.
    """

    def default(self, o):
        try:
            return super().default(o)
        except TypeError:
            # convert datetime to string format by default
            if isinstance(o, datetime):
                return o.isoformat()
            # convert object direct to string to avoid error in serialization
            return str(o)

File: libs/typescript/core/src/core/entities/span_event.ts
/**
 * MLflow Span Event Entities
 *
 * This module provides TypeScript implementations of MLflow span event entities,
 * compatible with OpenTelemetry for recording specific occurrences during span execution.
 */

/**
 * Type definition for attribute values that can be stored in span events.
 * Compatible with OpenTelemetry attribute value types.
 */
export type AttributeValue = string | number | boolean | string[] | number[] | boolean[];

/**
 * Parameters for creating a SpanEvent instance.
 */
export interface SpanEventParams {
  /** Name of the event */
  name: string;
  /**
   * The exact time the event occurred, measured in nanoseconds since epoch.
   * If not provided, the current time will be used.
   */
  timestamp?: bigint;
  /**
   * A collection of key-value pairs representing detailed attributes of the event,
   * such as exception stack traces or other contextual information.
   */
  attributes?: Record<string, AttributeValue>;
}

/**
 * Represents an event that records specific occurrences or moments in time
 * during a span, such as an exception being thrown. Compatible with OpenTelemetry.
 *
 * SpanEvents are used to capture important moments during span execution,
 * providing detailed context about what happened and when.
 *
 * @example
 * ```typescript
 * // Create a custom event
 * const event = new SpanEvent({
 *   name: 'user_action',
 *   attributes: {
 *     'action.type': 'click',
 *     'element.id': 'submit-button'
 *   }
 * });
 *
 * // Create an event from an exception
 * const errorEvent = SpanEvent.fromException(new Error('Something went wrong'));
 * ```
 */
export class SpanEvent {
  /** Name of the event */
  readonly name: string;

  /**
   * The exact time the event occurred, measured in nanosecond since epoch.
   * Defaults to current time if not provided during construction.
   */
  readonly timestamp: bigint;

  /**
   * A collection of key-value pairs representing detailed attributes of the event.
   * Attributes provide contextual information about the event.
   */
  readonly attributes: Record<string, AttributeValue>;

  /**
   * Creates a new SpanEvent instance.
   *
   * @param params - Event parameters including name, optional timestamp, and attributes
   *
   * @example
   * ```typescript
   * const event = new SpanEvent({
   *   name: 'database_query',
   *   attributes: {
   *     'db.statement': 'SELECT * FROM users',
   *     'db.duration_ms': 150
   *   }
   * });
   * ```
   */
  constructor(params: SpanEventParams) {
    this.name = params.name;
    this.timestamp = params.timestamp ?? this.getCurrentTimeNano();
    this.attributes = params.attributes ?? {};
  }

  /**
   * Creates a span event from an exception.
   *
   * This is a convenience method for creating events that represent exceptions
   * or errors that occurred during span execution. The event will include
   * standard exception attributes like message, type, and stack trace.
   *
   * @param exception - The exception to create an event from
   * @returns New SpanEvent instance representing the exception
   *
   * @example
   * ```typescript
   * try {
   *   // Some operation that might fail
   *   throw new Error('Database connection failed');
   * } catch (error) {
   *   const errorEvent = SpanEvent.fromException(error);
   *   span.addEvent(errorEvent);
   * }
   * ```
   */
  static fromException(exception: Error): SpanEvent {
    const stackTrace = this.getStackTrace(exception);

    return new SpanEvent({
      name: 'exception',
      attributes: {
        'exception.message': exception.message,
        'exception.type': exception.name,
        'exception.stacktrace': stackTrace
      }
    });
  }

  /**
   * Gets the stack trace from an error object.
   *
   * @param error - The error to extract stack trace from
   * @returns Stack trace as a string, or error representation if stack trace unavailable
   */
  private static getStackTrace(error: Error): string {
    try {
      return error.stack ?? String(error);
    } catch {
      // Fallback if stack trace extraction fails
      return String(error);
    }
  }

  /**
   * Convert this SpanEvent to JSON format
   * @returns JSON object representation of the span event
   */
  toJson(): Record<string, string | bigint | Record<string, AttributeValue>> {
    return {
      name: this.name,
      timestamp: this.timestamp,
      attributes: this.attributes
    };
  }

  /**
   * Gets the current time in nanoseconds since epoch.
   *
   * @returns Current timestamp in nanoseconds
   */
  private getCurrentTimeNano(): bigint {
    return BigInt(Date.now()) * BigInt(1e6);
  }
}
